libs/utils/src/logging.rs

```diff
@@ -326,41 +326,61 @@ impl std::fmt::Debug for SecretString {
 
 /// Logs a periodic message if a future is slow to complete.
 ///
+/// The callback is called at the same frequency.
+///
 /// This is performance-sensitive as it's used on the GetPage read path.
 ///
 /// TODO: consider upgrading this to a warning, but currently it fires too often.
 #[inline]
-pub async fn log_slow<O>(name: &str, threshold: Duration, f: impl Future<Output = O>) -> O {
-    // TODO: we unfortunately have to pin the future on the heap, since GetPage futures are huge and
-    // won't fit on the stack.
-    let mut f = Box::pin(f);
-
+pub async fn log_slow<F, O>(
+    name: &str,
+    threshold: Duration,
+    mut cb: impl FnMut(LogSlowCallback),
```

Comment:
Most callers won't need a callback. I'd split it out to a `log_slow_with` function, and have `log_slow` call through to it.